mila_datamodules/cli/prepare_torchvision.py

```diff
@@ -104,8 +139,83 @@ def __call__(
             archive_symlink.parent.mkdir(parents=True, exist_ok=True)
             archive_symlink.symlink_to(archive)
             print(f"Making link from {archive_symlink} -> {archive}")
-        # TODO: Check that nesting `runs_on_local_main_process_first` decorators isn't a problem.
-        return super().__call__(root, *constructor_args, **constructor_kwargs)
+
+        return str(root)
+
+
+class ExtractArchives(AbstractPrepareVisionDataset[VD, P]):
+    """Reorganize datasets' files in the `root` directory."""
+
+    def __init__(
+        self,
+        archives: dict[str, str | Path] = dict(),
+    ):
+        """
+        Parameters
+        ----------
+
+        - archives:
+            A list of pair of an archive and a destination's path where the
+            archive should be extracted. The destination's path should be
+            relative.
+        """
+        self.archives = [(glob, Path(path)) for glob, path in archives]
+
+    @runs_on_local_main_process_first
+    def __call__(
+        self,
+        root: str | Path = SLURM_TMPDIR / "datasets",
+        *dataset_args: P.args,
+        **dataset_kwargs: P.kwargs,
+    ) -> str:
+        for archive, dest in self.reorganize_files:
+            assert not dest.is_absolute()
+            if archive.suffix == ".zip":
+                with ZipFile(archive) as zf:
+                    zf.extractall(str(dest))
+            else:
+                raise ValueError("Unsupported archive type")
+
+        return str(root)
+
+
+class MoveFiles(AbstractPrepareVisionDataset[VD, P]):
+    """Reorganize datasets' files in the `root` directory."""
+
+    def __init__(
+        self,
+        files: list[str, str | Path] = dict(),
+    ):
+        """
+        Parameters
+        ----------
+
+        - files:
+            A list of pair of an archive and a destination's path where the result
+            should be moved and replace. If the destination path's leaf is "*",
+            the destination's parent will be used to hold the file. If not, the
+            destination will be used as the target for the move. The move are
+            executed in sequence. The destination's path should be relative.
```

Comment:
I'm trying to make this a bit easier to understand, but it's still a bit unclear.
```suggestion
            A mapping from file names to a destination path (relative to a
            'root' dir) where the file should be moved.
            If the destination path ends with "/*" (e.g. "data/*"), the files will
            be moved to the parent folder. Otherwise, the file is moved to
            the destination path.
            Files are moved in sequence. The destination's path need to be relative.
```

Reply:
I'm used to reference a `dict` with _mapping_. You would use _mapping_ as well with a `list` of `tuples`?

Reply:
But in this case, it's a mapping, right? not a list of tuples... Or if it is a list of tuples, then why not use a mapping instead?

Reply:
(since the first item is a string, I don't see why this can't be a dict mapping from `something` -> `something_renamed`)

Reply:
That is true but my brain in that case tends to still reserve some keywords to the type and explain in plain words how the object should be constructed or will be used. Your call :)